middleware/recover.go

```diff
@@ -0,0 +1,59 @@
+package middleware
+
+import (
+	"bytes"
+	"errors"
+	"net/http"
+	"runtime"
+	"strings"
+
+	"github.com/go-logr/logr"
+)
+
+// Recover from panics and logs the error.
+func Recover(logger logr.Logger) func(http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			defer func() {
+				if rec := recover(); rec != nil {
+					// Don't recover if the request is aborted, as this would
+					// prevent the client from detecting the error.
+					if rec == http.ErrAbortHandler {
+						panic(rec)
+					}
+
+					// Prepare the error message and log it.
+					b := strings.Builder{}
+					switch x := rec.(type) {
+					case string:
+						b.WriteString("panic: ")
+						b.WriteString(x)
+					case error:
+						b.WriteString("panic: ")
+						b.WriteString(x.Error())
+					default:
+						b.WriteString("unknown panic")
+					}
+					const size = 64 << 10 // 64KB
+					buf := make([]byte, size)
+					buf = buf[:runtime.Stack(buf, false)]
+					lines := bytes.Split(buf, []byte{'\n'})
+					if len(lines) > 3 {
```

Comment:
Why `len(lines) > 3`? Is three lines some sort of magic number in a stack trace?

Reply:
I believe that it's trying to skip two levels of stack that would create noise for the user, i.e. the call to the middleware itself and the runtime, but I'm glad you asked because this is very old code that I've been carrying over from past projects and I should investigate whether there is a better solution available!